### PR TITLE
Memoize some CngKey standard properties

### LINKED ISSUE #89821: Consider memoizing more CNG properties on CngKey that can't change after the key is finalized
#89599 shows that accessing some CNG properties are relatively expensive to the cryptographic operation itself. That pull request memoized the result for KeySize so that reading the KeySize property doesn't incur the overhead of reading ncrypt properties, which requires an RPC call.

We should consider what other properties can be memoized for performance. https://github.com/dotnet/runtime/pull/89819 showed that `AlgorithmGroup` is likely another candidate.

### PR DESCRIPTION
**Pull request by vcsjones**
This memoizes some CngKey properties similar to what we did for KeySize in https://github.com/dotnet/runtime/pull/89599 for performance. Accessing CNG key properties by Win32 have non-trivial performance characteristics since it results in an LRPC call. These properties are set during key creation time and cannot be changed after-the-fact, so storing them is reasonable.

Closes #89821

### CODE REVIEW

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs [hunk 1]
```diff
@@ -40,14 +49,22 @@ public CngAlgorithm Algorithm
...
+
+                    if (algorithmGroup is not null)
+                    {
+                        _cachedAlgorithmGroup = new CngAlgorithmGroup(algorithmGroup);
+                    }
+
+                    _hasCachedAlgorithmGroup = true;
+                }
+
+                return _cachedAlgorithmGroup;
```
Comment (stephentoub) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs:67]: Is CngKey supported for multithreaded reads? If so, this needs some volatile love.

Comment (vcsjones) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs:67]: Hm, reads, yeah. Good point.

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs [hunk 2]
```diff
@@ -279,10 +296,19 @@ public IntPtr ParentWindowHandle
         {
             get
             {
-                string? provider = _providerHandle.GetPropertyAsString(ProviderPropertyName.Name, CngPropertyOptions.None);
-                if (provider == null)
-                    return null;
-                return new CngProvider(provider);
+                if (!_hasCachedProvider || _keyHandle.IsClosed)
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs:299]: The fact that we need to check _keyHandle.IsClosed makes me think "obviously this should use a helper".  I think that helper would look like

```C#
private T GetCachedStringProperty<T>(ref bool hasValue, ref T field, string propertyName, Func<string?, T> resolver)
{
    if (!hasValue || _keyHandle.IsClosed)
    {
        string? value = _providerHandle.GetPropertyAsString(propertyName, CngPropertyOptions.None);
        T resolved = resolver(value);
        field = resolved;

        // Does the volatileness of the field flow through ref, or does this need to be explicit?
        Volatile.Write(ref hasValue, true);
    }

    return field;
}
```

and then the calls are

```C#
public CngProvider? Provider =>
    GetCachedStringProperty(ref _hasCachedProvider, ref _cachedProvider, ProviderPropertyName.Name, s => s is null ? null : new CngProvider(s));
```

which... is not necessarily better looking.

But the good news is it showed me a bug in this code :smile:

Comment (bartonjs) [inline src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs:299]:
```suggestion
                if (!_hasCachedProvider || _providerHandle.IsClosed)
```

The closed handle check is not using the same handle that's about to be queried.

**Review by bartonjs — APPROVED** [review]

File: src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs (hunk 2 above)
Comment (vcsjones) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs]: That would only be usable for 2 out of the 3 properties being memoized, or we could use a bool field unnecessarily. Plus the fact that you had to ask yourself a question about volatile kind of puts in the "not worth it" camp.

Comment (vcsjones) [reply src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/CngKey.StandardProperties.cs]: It doesn't matter in this particular case, but, agree it looks weird, so fixed.

**Pull request merged**